src/PortalGunHooks.js

```diff
@@ -0,0 +1,86 @@
+// @flow
+import * as React from 'react';
+
+class Store {
+  static id = 100000;
+
+  listeners = [];
+  idsToNodes = new Map();
+
+  addNode(id, node) {
+    this.idsToNodes.set(id, node);
+  }
+
+  deleteNode(id) {
+    this.idsToNodes.delete(id);
+    this.queueUpdate();
+  }
+
+  addListener(handler) {
+    this.listeners.push(handler);
+  }
+
+  removeListener(handler) {
+    this.listeners.splice(this.listeners.indexOf(handler), 1);
+  }
+
+  queueUpdate() {
+    setTimeout(() => this.listeners.forEach(listener => listener()), 0);
+  }
+}
+
+const useId = () => {
+  const ref = React.useRef();
+  if (!ref.current) {
+    Store.id += 1;
+    ref.current = Store.id;
+  }
+  return ref.current;
+};
+
+function SourceCmp({ children, store }) {
+  const id = useId();
+
+  React.useEffect(() => {
+    store.addNode(id, children);
```

Comment:
This actually needs to happen outside of the useEffect, or else there's a delay before it can render.